### ISSUE TITLE
Encode and Decode error event

### ISSUE THREAD

**Issue by drkron** [w3c/webrtc-pc#2219]
I'm looking for a way to detect decoding errors (both in the sense of failure to decode a frame and also in the sense that the decoder is falling behind due to scarce resources.

My suggestion is to add error events for decoding and encoding errors.

This was also proposed two years ago in #1323, where the consensus was that getStats() would be good enough.

The primary target for me is to use this to detect when the decoder is falling behind.
A concern I have with using getStats() is that it would require the user to have some custom heuristic code to detect this. Since this would be based on a rather limited set of information it would be noisy and it also requires the user to constantly pull getStats() for information. I believe that a signaling event that is fired from within WebRTC should be possible to make more accurate and robust.

Another problem is that some HW encoder/decoder uses a pipeline so that there's always a one-frame delay. In this case a high encode/decode time does not mean that it's falling behind. In such a case a detection in WebRTC could look at how queues are building up.

**Comment by aboba** [comment]
There are several types of errors that can occur.  A device (such as a webcam) can have a built-in encoder.  Sometimes the device will successfully initialize, but later on there will be transient encoding errors, or perhaps encoding will stop.   Then there are transient or fatal errors in hardware or software decoders/encoders.  If encoding/decoding totally stops, then the application needs to known this. 
There is a  DOMException with a name of `EncodingError` that we could use.

**Comment by aboba** [comment]
@drkron Would you be available to present this issue (and potential solutions) at the July 2 Virtual Interim?

**Comment by drkron** [comment]
> @drkron Would you be available to present this issue (and potential solutions) at the July 2 Virtual Interim?

I'm on vacation but I will try to attend the meeting.

**Comment by henbos** [comment]
> Is there an agenda for the meeting, that is, is it decided in which order all issues are discussed?

Yes, see the [slides](https://docs.google.com/presentation/d/1xcvf0udNeSH7s1FOY7RRqr1dEFvokZjn-MZPjwy3iXQ/edit?usp=sharing). I gave you write-access with your google account

**Comment by aboba** [comment]
Note that there is an RTCError defined for a hardware encoder error in Section 5.2 (used by `setParameters`):
  
"If an error occurred due to a hardware encoder not supporting parameters, reject `p` with a newly created `RTCError` whose `errorDetail` is set to "hardware-encoder-error"

**Comment by drkron** [comment]
I'm looking into using the existing stats, framesReceived and framesDecoded, to detect if the decoder is too slow. From my initial testing it seems as taking the difference gives a quite good indication of the current decoding status.
So for now I don't see the immediate need for this feature.
